### ISSUE TITLE
Resolve libc::time_t issues

### ISSUE THREAD

**Issue by ridiculousfish** [fish-shell/fish-shell#10634]
See https://github.com/rust-lang/libc/issues/1848

When compiling on musl, fish emits warnings:

```
use of deprecated type alias `libc::time_t`: This type is changed to 64-bit in musl 1.2.0, we'll follow that change in the future release. See #1848 for more info.
```

[A comment from that issue]:(https://github.com/rust-lang/libc/issues/1848#issuecomment-1259700917)

```
The only portable way to do this is to avoid any use of time_t, struct timeval, struct timespec, ino_t, blkcnt_t, as well as any types derived from these, such as struct stat, struct dirent, or struct rlimit when interfacing with C/C++ code.
```

which seems concerning.

**Comment by krobelus** [comment]
changed our uses but the stdlib is still broken (for example `std::thread::sleep`)

**Comment by faho** [comment]
I'm not entirely sure we're understanding this correctly.

As I understand it from reading https://github.com/ericonr/rust-time64, there are two problems. One's y2038 safety, which I assume we're fine with punting for when rust updates.

The other is only a problem if we passed time_t to a C library other than libc.

> This isn't an issue for pure Rust projects, because the libc crate also hardcodes the function names, so it links to musl's time32 compatibility functions, instead of the modern time64 ones. The only issue here is that the resulting binaries aren't safe for y2038.

And as far as I understand we do not do that (our only other C dependency is pcre, and we don't give that a time_t), and so I am unsure if there is anything we actually need to do here.

Is this actually necessary?

**Comment by krobelus** [comment]
I think the problem is corruption when `std::mem::size_of::<libc::time_t>()` doesn't match `sizeof(struct time_t)` in a glibc routine.
The example is musl (which always uses 64 bit `time_t`) on a 32 bit system.

Given

    pub fn sleep(dur: Duration) {
        let mut ts = libc::timespec {
            tv_sec: cmp::min(libc::time_t::MAX as u64, secs) as libc::time_t,
            tv_nsec: nsecs,
        };
        libc::nanosleep(&ts, std::ptr::null_mut())
    }

The definitions on the Rust side are effectively

    struct timespec {
        tv_sec: u32,
        tv_nsec: u32,
    }
    fn nanosleep(rqtp: *const timespec, rmtp: *mut timespec) -> ::c_int;

and musl uses

    struct timespec {
        uint64_t tv_sec;
        uint64_t tv_nsec;
    }

So reading `tv_sec` yields garbage.

There has been some work on fixing this for all of libc: https://github.com/rust-lang/libc/pull/3175

For now we could probably add a compile error for this scenario and (though it will surely crash soon enough), marking it as unsupported.

**Comment by faho** [comment]
> I think the problem is corruption when std::mem::size_of::<libc::time_t>() doesn't match 
sizeof(struct time_t) in a glibc routine.

I believe you meant "libc" there, this is only about musl. Glibc uses a different way of making time_t 64-bit and is, as far as I know, entirely unaffected. The rust warning is also specifically about musl and only appears when compiling for musl.

What I want to know is if this:

>  it links to musl's time32 compatibility functions, instead of the modern time64 ones. 

is correct. Because if it is we don't need to do anything - we use the 32-bit versions for now, and since we don't pass them anywhere except for libc (where we would use the 32-bit versions of these functions), I don't believe there is a problem?

You can also see https://github.com/bminor/musl/commit/38143339646a4ccce8afe298c34467767c899f51, which states:

> this commit preserves ABI fully for existing interface boundaries
between libc and libc consumers (applications or libraries), by
retaining existing symbol names for the legacy 32-bit interfaces

So I believe that rust is simply using those legacy interfaces for now, and the issue appears when you link against *another* C library (in addition to libc) that does not use them.

**Comment by faho** [comment]
Okay, I have built a bunch of releases with musl, including i686 ones. Even after reverting all the libc.c changes (the ones that use more rust std are fine and should stay), I can not find a problem except for it not being 2038-safe (do `touch -d '11/11/2039' somefile` and `path mtime somefile` and it will give you a negative value in an i686 build - but this is also the case *with* the libc.c changes).

I do not believe this is an issue *for us*. I believe the problem is that, if you're linking against other C libraries and pass time_t to them, you need to arrange for them to be built against musl < 1.2.0 right now, and you need to build them against >= 1.2.0 in future. That does not apply to us.

I would close it and revert the libc.c changes - 55196ee2a0430d920ea7a2c89a6e322615f78334, 4992f8896633fb8ca8d89e09f02330cd49395485, 46c8ba2c9fec77195091ddcf7ee0bb3d9a6e5f54, 3a9b4149da7d44b8648702f17d9e9eef651e56f9, 5f9e9cbe741025231acfb24dc900433e1c6738ac.

**Comment by krobelus** [comment]
yeah that sounds right. We don't have to worry about preprocessor macros like `_TIME_BITS`.
Once `libc` provides 64-bit versions to replace the deprecated word-sized ones, then we can switch to those.
It's a bit strange to deprecate something without having an alternative ready.
